File: week_1/code/invert_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        """
        Given the root of a binary tree, invert the tree, and return its root.
        """
        if not root:
            return None
        
        # Swap and recurse
        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        
        return root

# Helper to print tree (Level Order) for testing
def printTree(root):
    if not root:
        print("[]")
        return
    queue = [root]
    res = []
    while queue:
        node = queue.pop(0)
        if node:
            res.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            res.append("null")
    # Clean up trailing nulls
    while res and res[-1] == "null":
        res.pop()
    print("[" + ",".join(res) + "]")

# Helper to build tree from level order list
def buildTree(arr):
    if not arr: return None
    root = TreeNode(arr[0])
    queue = [root]
    i = 1
    while queue and i < len(arr):
        node = queue.pop(0)
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root

# Test cases
if __name__ == "__main__":
    sol = Solution()
    
    # Test 1: [4,2,7,1,3,6,9]
    root1 = buildTree([4,2,7,1,3,6,9])
    print("Original 1:")
    printTree(root1)
    inverted1 = sol.invertTree(root1)
    print("Inverted 1 (Expected: [4,7,2,9,6,3,1]):")
    printTree(inverted1)

    # Test 2: [2,1,3]
    root2 = buildTree([2,1,3])
    print("Original 2:")
    printTree(root2)
    inverted2 = sol.invertTree(root2)
    print("Inverted 2 (Expected: [2,3,1]):")
    printTree(inverted2)

File: week_1/code/invert_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function (root) {
    if (!root) return null;

    let temp = root.left;
    root.left = invertTree(root.right);
    root.right = invertTree(temp);

    return root;
};

// Helper to print tree (Level Order)
function printTree(root) {
    if (!root) {
        console.log("[]");
        return;
    }
    let queue = [root];
    let res = [];
    while (queue.length > 0) {
        let node = queue.shift();
        if (node) {
            res.push(node.val);
            queue.push(node.left);
            queue.push(node.right);
        } else {
            res.push("null");
        }
    }
    // Clean up trailing nulls
    while (res.length > 0 && res[res.length - 1] === "null") {
        res.pop();
    }
    console.log("[" + res.join(",") + "]");
}

// Helper to build tree from array
function buildTree(arr) {
    if (!arr.length) return null;
    let root = new TreeNode(arr[0]);
    let queue = [root];
    let i = 1;
    while (queue.length > 0 && i < arr.length) {
        let node = queue.shift();
        if (i < arr.length && arr[i] !== null) {
            node.left = new TreeNode(arr[i]);
            queue.push(node.left);
        }
        i++;
        if (i < arr.length && arr[i] !== null) {
            node.right = new TreeNode(arr[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

// Test cases
// Test 1: [4,2,7,1,3,6,9]
let root1 = buildTree([4, 2, 7, 1, 3, 6, 9]);
console.log("Original 1:");
printTree(root1);
let inverted1 = invertTree(root1);
console.log("Inverted 1 (Expected: [4,7,2,9,6,3,1]):");
printTree(inverted1);

// Test 2: [2,1,3]
let root2 = buildTree([2, 1, 3]);
console.log("Original 2:");
printTree(root2);
let inverted2 = invertTree(root2);
console.log("Inverted 2 (Expected: [2,3,1]):");
printTree(inverted2);
